packages/firestore/src/local/web_storage.ts

```diff
@@ -0,0 +1,146 @@
+/**
+ * Copyright 2018 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Code, FirestoreError } from '../util/error';
+import { VisibilityState } from '../core/types';
+import { assert } from '../util/assert';
+import { AsyncQueue } from '../util/async_queue';
+import { debug } from '../util/log';
+
+/**
+ * Refresh the contents of WebStorage every four seconds.
+ */
+const WEB_STORAGE_REFRESH_INTERVAL_MS: number = 4000;
+
+// Prefix keys used in WebStorage.
+const VISIBILITY_PREFIX = 'visibility';
+
+const LOG_TAG = 'WebStorage';
+
+/**
+ * WebStorage of the Firestore client. Firestore uses WebStorage for cross-tab
+ * notifications and to persist the metadata state of each tab. WebStorage is
+ * used to perform leader election and to inform other tabs about changes in the
+ * IndexedDB-backed persistence layer.
+ *
+ * WebStorage is only used in Persistence-enabled Firestore instances. If you
+ * are not using Persistence, consider using the `NoOpWebStorage` class.
+ */
+export interface WebStorage {
+  setVisibility(visibilityState: VisibilityState): void;
+  start(): void;
+  shutdown(): void;
+}
+
+/**
+ * `PersistedWebStorage` uses Local Storage as the backing store for the
+ * WebStorage class.
+ *
+ * Once started, PersistedWebStorage will rewrite its contents to Local Storage
+ * every four seconds. Other clients may disregard its state after five seconds
+ * of inactivity.
+ */
+export class PersistedWebStorage implements WebStorage {
```

Comment:
If IndexedDB listeners became available someday and we wanted to use either those or this one maybe it would be better to name this after how it's implemented?

Is LocalStorageTabNotificationChannel too much?

Reply:
I'll get back to you 2025, but for now I named it`LocalStorageNotificationChannel`.